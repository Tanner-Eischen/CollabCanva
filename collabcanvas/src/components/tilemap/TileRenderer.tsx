/**
 * TileRenderer Component
 * Efficiently renders tiles with viewport culling
 * Supports both sprite and colored tile rendering
 */

import { useMemo } from 'react'
import { Layer, Rect } from 'react-konva'
import type { TileData } from '../../types/tilemap'
import { coordToKey } from '../../types/tilemap'
import { getTilePath, hasSpriteAsset } from '../../constants/tilemapDefaults'
import SpriteTile from '../canvas/SpriteTile'

interface TileRendererProps {
  tiles: Map<string, TileData>
  tileSize: number
  viewportX: number
  viewportY: number
  viewportWidth: number
  viewportHeight: number
  previewTile?: { x: number; y: number; tile: TileData } | null
  showPreview?: boolean
}

/**
 * Tile renderer with viewport culling
 * Only renders tiles visible in the current viewport for performance
 */
export default function TileRenderer({
  tiles,
  tileSize,
  viewportX,
  viewportY,
  viewportWidth,
  viewportHeight,
  previewTile,
  showPreview = true,
}: TileRendererProps) {
  // Calculate visible tiles with viewport culling
  const visibleTiles = useMemo(() => {
    const visible: Array<{ key: string; x: number; y: number; tile: TileData }> = []
    
    // Calculate visible tile range with some padding
    const padding = 2 // Extra tiles outside viewport
    const startX = Math.max(0, Math.floor(viewportX / tileSize) - padding)
    const endX = Math.ceil((viewportX + viewportWidth) / tileSize) + padding
    const startY = Math.max(0, Math.floor(viewportY / tileSize) - padding)
    const endY = Math.ceil((viewportY + viewportHeight) / tileSize) + padding
    
    // Iterate through tiles and filter visible ones
    tiles.forEach((tile, key) => {
      const [xStr, yStr] = key.split('_')
      const x = parseInt(xStr, 10)
      const y = parseInt(yStr, 10)
      
      // Check if tile is in visible range
      if (x >= startX && x <= endX && y >= startY && y <= endY) {
        visible.push({ key, x, y, tile })
      }
    })
    
    return visible
  }, [tiles, tileSize, viewportX, viewportY, viewportWidth, viewportHeight])
  
  return (
    <Layer
      listening={false}
      perfectDrawEnabled={false}
    >
      {/* Render visible tiles */}
      {visibleTiles.map(({ key, x, y, tile }) => {
        const hasSprite = hasSpriteAsset(tile.type)
        
        // Render sprite tile if available, otherwise colored rect
        if (hasSprite && tile.variant !== undefined) {
          // Clamp variant to valid range (0-8) to handle old data
          const clampedVariant = Math.max(0, Math.min(8, tile.variant))
          const tilePath = getTilePath(tile.type, clampedVariant)
          return (
            <SpriteTile
              key={key}
              x={x * tileSize}
              y={y * tileSize}
              tileSize={tileSize}
              tilePath={tilePath}
              color={tile.color}
              opacity={1}
            />
          )
        }
        
        // Fallback to colored rectangle (backwards compatible)
        return (
          <Rect
            key={key}
            x={x * tileSize}
            y={y * tileSize}
            width={tileSize}
            height={tileSize}
            fill={tile.color}
            stroke="rgba(0, 0, 0, 0.1)"
            strokeWidth={1}
            listening={false}
            perfectDrawEnabled={false}
          />
        )
      })}
      
      {/* Render preview tile (ghost) */}
      {showPreview && previewTile && (() => {
        const hasSprite = hasSpriteAsset(previewTile.tile.type)
        
        // Render sprite preview if available
        if (hasSprite && previewTile.tile.variant !== undefined) {
          // Clamp variant to valid range (0-8)
          const clampedVariant = Math.max(0, Math.min(8, previewTile.tile.variant))
          const tilePath = getTilePath(previewTile.tile.type, clampedVariant)
          return (
            <SpriteTile
              key="preview"
              x={previewTile.x * tileSize}
              y={previewTile.y * tileSize}
              tileSize={tileSize}
              tilePath={tilePath}
              color={previewTile.tile.color}
              opacity={0.5}
            />
          )
        }
        
        // Fallback to colored preview
        return (
          <Rect
            key="preview"
            x={previewTile.x * tileSize}
            y={previewTile.y * tileSize}
            width={tileSize}
            height={tileSize}
            fill={previewTile.tile.color}
            opacity={0.5}
            stroke="#3b82f6"
            strokeWidth={2}
            dash={[4, 4]}
            listening={false}
            perfectDrawEnabled={false}
          />
        )
      })()}
    </Layer>
  )
}

