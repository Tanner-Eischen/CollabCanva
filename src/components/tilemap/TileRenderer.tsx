/**
 * TileRenderer Component
 * Efficiently renders tiles with viewport culling
 * Supports both sprite and colored tile rendering
 * Supports multi-layer rendering with parallax
 * Performance optimized with React.memo and Konva settings
 */

import React, { useEffect, useMemo, useState } from 'react'
import { Layer, Rect } from 'react-konva'
import type { TileData } from '../../types/tilemap'
import type { TileLayerMeta } from '../../types/tileLayer'
import { applyParallax } from '../../types/tileLayer'
import SpriteTile from '../canvas/SpriteTile'
import AnimatedTile from '../canvas/AnimatedTile'
import { tilesetRegistry } from '../../services/tilemap/tilesetRegistry'

const NEUTRAL_FALLBACK_COLOR = '#9ca3af'

interface TileRendererProps {
  tiles: Map<string, TileData>
  tileSize: number
  viewportX: number
  viewportY: number
  viewportWidth: number
  viewportHeight: number
  previewTile?: { x: number; y: number; tile: TileData } | null
  showPreview?: boolean
  layer?: TileLayerMeta // Optional layer metadata for parallax and opacity
}

/**
 * Tile renderer with viewport culling
 * Only renders tiles visible in the current viewport for performance
 * Memoized to prevent unnecessary re-renders
 */
function TileRenderer({
  tiles,
  tileSize,
  viewportX,
  viewportY,
  viewportWidth,
  viewportHeight,
  previewTile,
  showPreview = true,
  layer,
}: TileRendererProps) {
  // Apply parallax offset to viewport if layer has parallax config
  const { parallaxOffsetX, parallaxOffsetY, layerOpacity } = useMemo(() => {
    if (layer?.parallax) {
      return {
        parallaxOffsetX: viewportX - applyParallax(viewportX, layer.parallax.x),
        parallaxOffsetY: viewportY - applyParallax(viewportY, layer.parallax.y),
        layerOpacity: layer.opacity ?? 1,
      }
    }
    return {
      parallaxOffsetX: 0,
      parallaxOffsetY: 0,
      layerOpacity: layer?.opacity ?? 1,
    }
  }, [layer, viewportX, viewportY])

  // Calculate visible tiles with viewport culling (accounting for parallax)
  const visibleTiles = useMemo(() => {
    const visible: Array<{ key: string; x: number; y: number; tile: TileData }> = []
    
    // Adjust viewport for parallax when calculating culling
    const effectiveViewportX = viewportX - parallaxOffsetX
    const effectiveViewportY = viewportY - parallaxOffsetY
    
    // Calculate visible tile range with some padding
    const padding = 2 // Extra tiles outside viewport
    const startX = Math.max(0, Math.floor(effectiveViewportX / tileSize) - padding)
    const endX = Math.ceil((effectiveViewportX + viewportWidth) / tileSize) + padding
    const startY = Math.max(0, Math.floor(effectiveViewportY / tileSize) - padding)
    const endY = Math.ceil((effectiveViewportY + viewportHeight) / tileSize) + padding
    
    // Iterate through tiles and filter visible ones
    tiles.forEach((tile, key) => {
      const [xStr, yStr] = key.split('_')
      const x = parseInt(xStr, 10)
      const y = parseInt(yStr, 10)
      
      // Check if tile is in visible range
      if (x >= startX && x <= endX && y >= startY && y <= endY) {
        visible.push({ key, x, y, tile })
      }
    })
    
    return visible
  }, [tiles, tileSize, viewportX, viewportY, viewportWidth, viewportHeight, parallaxOffsetX, parallaxOffsetY])
  
  return (
    <Layer
      listening={false}
      perfectDrawEnabled={false}
      hitGraphEnabled={false} // Disable hit detection for better performance
      imageSmoothingEnabled={false} // Faster rendering for pixel art
      opacity={layerOpacity}
      // Apply parallax offset to the entire layer
      x={parallaxOffsetX}
      y={parallaxOffsetY}
    >
      {/* Render visible tiles */}
      {visibleTiles.map(({ key, x, y, tile }) => {
        // Render animated tile if animation ID present
        if (tile.animationId) {
          return (
            <AnimatedTile
              key={key}
              x={x * tileSize}
              y={y * tileSize}
              tileSize={tileSize}
              animationId={tile.animationId}
              color={tile.color}
              opacity={1}
              isPlaying={true}
            />
          )
        }

        return (
          <TileSpriteOrRect
            key={key}
            x={x * tileSize}
            y={y * tileSize}
            tileSize={tileSize}
            tile={tile}
            opacity={1}
          />
        )
      })}
      
      {/* Render preview tile (ghost) */}
      {showPreview && previewTile && (() => {
        // Render animated preview if animation ID present
        if (previewTile.tile.animationId) {
          return (
            <AnimatedTile
              key="preview"
              x={previewTile.x * tileSize}
              y={previewTile.y * tileSize}
              tileSize={tileSize}
              animationId={previewTile.tile.animationId}
              color={previewTile.tile.color}
              opacity={0.5}
              isPlaying={true}
            />
          )
        }
        
        return (
          <TileSpriteOrRect
            key="preview"
            x={previewTile.x * tileSize}
            y={previewTile.y * tileSize}
            tileSize={tileSize}
            tile={previewTile.tile}
            opacity={0.5}
            isPreview
          />
        )
      })()}
    </Layer>
  )
}

interface TileSpriteOrRectProps {
  tile: TileData
  tileSize: number
  x: number
  y: number
  opacity: number
  isPreview?: boolean
}

function TileSpriteOrRect({ tile, tileSize, x, y, opacity, isPreview = false }: TileSpriteOrRectProps) {
  const [hasSprite, setHasSprite] = useState(false)
  const [tilePath, setTilePath] = useState<string | null>(null)
  const [fillColor, setFillColor] = useState<string>(tile.color ?? NEUTRAL_FALLBACK_COLOR)

  useEffect(() => {
    let isMounted = true

    setHasSprite(false)
    setTilePath(null)
    setFillColor(tile.color ?? NEUTRAL_FALLBACK_COLOR)

    const loadSpriteData = async () => {
      try {
        const [spriteAvailable, registryColor] = await Promise.all([
          tilesetRegistry.hasSprite(tile.type),
          tilesetRegistry.getTerrainColor(tile.type),
        ])

        if (!isMounted) {
          return
        }

        const resolvedFill = registryColor ?? tile.color ?? NEUTRAL_FALLBACK_COLOR
        setFillColor(resolvedFill)

        if (!spriteAvailable) {
          return
        }

        const path = await tilesetRegistry.getTilePath(tile.type, tile.variant ?? 4)

        if (!isMounted) {
          return
        }

        if (path) {
          setTilePath(path)
          setHasSprite(true)
        }
      } catch (_error) {
        if (!isMounted) {
          return
        }
        setHasSprite(false)
        setTilePath(null)
      }
    }

    void loadSpriteData()

    return () => {
      isMounted = false
    }
  }, [tile.type, tile.variant, tile.color])

  if (hasSprite && tilePath) {
    return (
      <SpriteTile
        x={x}
        y={y}
        tileSize={tileSize}
        tilePath={tilePath}
        color={tile.color ?? fillColor}
        opacity={opacity}
      />
    )
  }

  return (
    <Rect
      x={x}
      y={y}
      width={tileSize}
      height={tileSize}
      fill={fillColor}
      opacity={opacity}
      stroke={isPreview ? '#3b82f6' : 'rgba(0, 0, 0, 0.1)'}
      strokeWidth={isPreview ? 2 : 1}
      dash={isPreview ? [4, 4] : undefined}
      listening={false}
      perfectDrawEnabled={false}
    />
  )
}

/**
 * Export memoized version to prevent unnecessary re-renders
 * Only re-render if tiles, viewport, or layer changes
 */
export default React.memo(TileRenderer, (prevProps, nextProps) => {
  // Only re-render if these specific props changed
  return (
    prevProps.tiles === nextProps.tiles &&
    Math.floor(prevProps.viewportX / prevProps.tileSize) === Math.floor(nextProps.viewportX / nextProps.tileSize) &&
    Math.floor(prevProps.viewportY / prevProps.tileSize) === Math.floor(nextProps.viewportY / nextProps.tileSize) &&
    prevProps.viewportWidth === nextProps.viewportWidth &&
    prevProps.viewportHeight === nextProps.viewportHeight &&
    prevProps.layer?.id === nextProps.layer?.id &&
    prevProps.layer?.visible === nextProps.layer?.visible &&
    prevProps.layer?.opacity === nextProps.layer?.opacity &&
    prevProps.previewTile === nextProps.previewTile
  )
})

