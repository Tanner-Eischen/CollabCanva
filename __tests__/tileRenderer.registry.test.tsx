import React from 'react'
import { render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import type { TileData } from '../src/types/tilemap'

vi.mock('../src/services/firebase', () => ({
  firestore: {},
  db: {},
}))

vi.mock('react-konva', () => {
  const React = require('react')
  return {
    Layer: ({ children, ...props }: { children: React.ReactNode }) => {
      const {
        listening: _listening,
        perfectDrawEnabled: _perfectDrawEnabled,
        hitGraphEnabled: _hitGraphEnabled,
        imageSmoothingEnabled: _imageSmoothingEnabled,
        ...rest
      } = props as Record<string, unknown>
      return (
        <div data-component="Layer" {...rest}>
          {children}
        </div>
      )
    },
    Rect: ({ children, ...props }: { children?: React.ReactNode }) => {
      const {
        listening: _listening,
        perfectDrawEnabled: _perfectDrawEnabled,
        hitGraphEnabled: _hitGraphEnabled,
        imageSmoothingEnabled: _imageSmoothingEnabled,
        ...rest
      } = props as Record<string, unknown>
      return (
        <div data-component="Rect" {...rest}>
          {children}
        </div>
      )
    },
  }
})

vi.mock('../src/components/canvas/SpriteTile', () => {
  const React = require('react')
  return {
    __esModule: true,
    default: ({ tilePath, x, y, tileSize, opacity }: Record<string, unknown>) => (
      <div
        data-testid="sprite-tile"
        data-path={tilePath as string}
        data-x={x as number}
        data-y={y as number}
        data-size={tileSize as number}
        data-opacity={opacity as number}
      />
    ),
  }
})

vi.mock('../src/components/canvas/AnimatedTile', () => ({
  __esModule: true,
  default: (props: Record<string, unknown>) => <div data-testid="animated-tile" {...props} />,
}))

import TileRenderer from '../src/components/tilemap/TileRenderer'

describe('TileRenderer integration with tileset registry', () => {
  it('renders sprite tiles for grass/dirt boundary using registry data', async () => {
    const tiles = new Map<string, TileData>([
      ['0_0', { type: 'grass', color: '#4ade80', variant: 4 }],
      ['1_0', { type: 'dirt', color: '#b45309', variant: 4 }],
    ])

    const { asFragment } = render(
      <TileRenderer
        tiles={tiles}
        tileSize={32}
        viewportX={0}
        viewportY={0}
        viewportWidth={64}
        viewportHeight={32}
      />
    )

    const spriteTiles = await screen.findAllByTestId('sprite-tile')
    expect(spriteTiles).toHaveLength(2)

    expect(spriteTiles.map((tile) => tile.getAttribute('data-path'))).toMatchInlineSnapshot(`
      [
        "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAAKklEQVR42mNwOl5MU8QwasGoBaMWjFowasGoBaMWjFowasGoBaMWDBULAGQg8Ey6liD6AAAAAElFTkSuQmCC",
        "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAAKklEQVR42mNY5MVBU8QwasGoBaMWjFowasGoBaMWjFowasGoBaMWDBULAM8A0C5B34lnAAAAAElFTkSuQmCC",
      ]
    `)

    expect(asFragment()).toMatchSnapshot()
  })
})
