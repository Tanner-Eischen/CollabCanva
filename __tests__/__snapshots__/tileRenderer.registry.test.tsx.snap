// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TileRenderer integration with tileset registry > renders sprite tiles for grass/dirt boundary using registry data 2`] = `
<DocumentFragment>
  <div
    data-component="Layer"
    opacity="1"
    x="0"
    y="0"
  >
    <div
      data-opacity="1"
      data-path="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAAKklEQVR42mNwOl5MU8QwasGoBaMWjFowasGoBaMWjFowasGoBaMWDBULAGQg8Ey6liD6AAAAAElFTkSuQmCC"
      data-size="32"
      data-testid="sprite-tile"
      data-x="0"
      data-y="0"
    />
    <div
      data-opacity="1"
      data-path="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAIAAAD8GO2jAAAAKklEQVR42mNY5MVBU8QwasGoBaMWjFowasGoBaMWjFowasGoBaMWDBULAM8A0C5B34lnAAAAAElFTkSuQmCC"
      data-size="32"
      data-testid="sprite-tile"
      data-x="32"
      data-y="0"
    />
  </div>
</DocumentFragment>
`;
